File: apps/server2/src/api/components/user/user_controller.py
from typing import Annotated

from dependency_injector.wiring import Provide, inject
from fastapi import APIRouter, Depends, Query, Request, status
from fastapi.encoders import jsonable_encoder
from fastapi.responses import JSONResponse

from api.components.user.user_mapper import UserMapper
from api.components.user.user_models import UserRequest, UserResponse
from api.components.user.user_service import UserService
from api.shared.api_error_response import APIErrorResponse
from container.container import Container
from services.api_pagination_service import APIPaginationData, APIPaginationService


class UserController(APIRouter):
    def __init__(
        self,
        prefix="/users",
        dependencies=[
            Depends(Provide[Container.user_service_provider]),
            Depends(Provide[Container.api_pagination_service_provider]),
        ],
    ):
        super().__init__(prefix=prefix, dependencies=dependencies)
        self.setup_routes()

    def setup_routes(self):
        @APIRouter.api_route(
            self,
            path="",
            methods=["POST"],
            tags=["users"],
            description="API endpoint used to create a new user",
            responses={
                status.HTTP_201_CREATED: {
                    "model": UserResponse,
                    "description": "Created",
                    "content": {
                        "application/json": {
                            "example": {
                                "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
                                "name": "name",
                                "email": "email@email.com",
                                "created_at": "XXXX-XX-XXTXX:XX:XX.XXXXXX",
                                "updated_at": "",
                            }
                        }
                    },
                },
                status.HTTP_422_UNPROCESSABLE_ENTITY: {
                    "model": APIErrorResponse,
                    "description": "Unprocessable Entity",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "",
                                "details": {"context": "", "cause": ""},
                                "isOperational": True,
                            }
                        }
                    },
                },
                status.HTTP_500_INTERNAL_SERVER_ERROR: {
                    "model": APIErrorResponse,
                    "description": "Internal Server Error",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "Internal Server Error",
                                "details": {"context": "", "cause": ""},
                                "isOperational": False,
                            }
                        }
                    },
                },
            },
        )
        @inject
        async def add_user(
            user_request: UserRequest,
            user_service: UserService = self.dependencies[0],
        ) -> JSONResponse:
            domain_user = UserMapper.to_domain(user_request)
            returned_user = await user_service.register_user(domain_user)
            user_response = UserMapper.to_response(returned_user)
            response = JSONResponse(
                content=jsonable_encoder(user_response),
                status_code=status.HTTP_201_CREATED,
            )
            return response

        @APIRouter.api_route(
            self,
            path="",
            methods=["GET"],
            tags=["users"],
            description="""
            API endpoint used to get users through page-based pagination schema.
            * @param page The number of the page. If isn't provided, it will be set to 1.
            * @param limit The number of records per page. If isn't provided, it will be set to 1.
            """,
            responses={
                status.HTTP_200_OK: {
                    "model": UserResponse,
                    "description": "OK",
                    "content": {
                        "application/json": {
                            "example": {
                                "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
                                "name": "name",
                                "email": "email@email.com",
                                "created_at": "XXXX-XX-XXTXX:XX:XX.XXXXXX",
                                "updated_at": "",
                            }
                        }
                    },
                },
                status.HTTP_500_INTERNAL_SERVER_ERROR: {
                    "model": APIErrorResponse,
                    "description": "Internal Server Error",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "Internal Server Error",
                                "details": {"context": "", "cause": ""},
                                "isOperational": False,
                            }
                        }
                    },
                },
            },
        )
        @inject
        async def fetch_paginated_users(
            request: Request,
            page: Annotated[int | None, Query()] = 1,
            limit: Annotated[int | None, Query()] = 1,
            user_service: UserService = self.dependencies[0],
            api_pagination_service: APIPaginationService = self.dependencies[1],
        ) -> JSONResponse:
            base_url = str(request.url)
            (returned_users, count) = await user_service.retrieve_and_count_users(
                page, limit
            )
            api_pagination_data = APIPaginationData(
                page=page, limit=limit, total_records=count, records=returned_users
            )
            api_pagination_response = (
                api_pagination_service.create_api_pagination_response(
                    base_url, api_pagination_data
                )
            )
            response = JSONResponse(
                content=jsonable_encoder(api_pagination_response),
                status_code=status.HTTP_200_OK,
            )
            return response

        @APIRouter.api_route(
            self,
            path="/{user_id}",
            methods=["GET"],
            tags=["users"],
            description="API endpoint used to get a user by its ID",
            responses={
                status.HTTP_200_OK: {
                    "model": UserResponse,
                    "description": "OK",
                    "content": {
                        "application/json": {
                            "example": {
                                "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
                                "name": "name",
                                "email": "email@email.com",
                                "created_at": "XXXX-XX-XXTXX:XX:XX.XXXXXX",
                                "updated_at": "",
                            }
                        }
                    },
                },
                status.HTTP_404_NOT_FOUND: {
                    "model": APIErrorResponse,
                    "description": "Unprocessable Entity",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "",
                                "details": {"context": "", "cause": ""},
                                "isOperational": True,
                            }
                        }
                    },
                },
                status.HTTP_500_INTERNAL_SERVER_ERROR: {
                    "model": APIErrorResponse,
                    "description": "Internal Server Error",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "Internal Server Error",
                                "details": {"context": "", "cause": ""},
                                "isOperational": False,
                            }
                        }
                    },
                },
            },
        )
        @inject
        async def fetch_user(
            user_id: str,
            user_service: UserService = self.dependencies[0],
        ) -> JSONResponse:
            returned_user = await user_service.retrieve_user(user_id)
            user_response = UserMapper.to_response(returned_user)
            response = JSONResponse(
                content=jsonable_encoder(user_response),
                status_code=status.HTTP_200_OK,
            )
            return response

        @APIRouter.api_route(
            self,
            path="/{user_id}",
            methods=["PUT"],
            tags=["users"],
            description="API endpoint used to update a user by its ID",
            responses={
                status.HTTP_200_OK: {
                    "model": UserResponse,
                    "description": "OK",
                    "content": {
                        "application/json": {
                            "example": {
                                "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
                                "name": "name",
                                "email": "email@email.com",
                                "created_at": "XXXX-XX-XXTXX:XX:XX.XXXXXX",
                                "updated_at": "",
                            }
                        }
                    },
                },
                status.HTTP_404_NOT_FOUND: {
                    "model": APIErrorResponse,
                    "description": "Not Found",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "",
                                "details": {"context": "", "cause": ""},
                                "isOperational": True,
                            }
                        }
                    },
                },
                status.HTTP_422_UNPROCESSABLE_ENTITY: {
                    "model": APIErrorResponse,
                    "description": "Unprocessable Entity",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "",
                                "details": {"context": "", "cause": ""},
                                "isOperational": True,
                            }
                        }
                    },
                },
                status.HTTP_500_INTERNAL_SERVER_ERROR: {
                    "model": APIErrorResponse,
                    "description": "Internal Server Error",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "Internal Server Error",
                                "details": {"context": "", "cause": ""},
                                "isOperational": False,
                            }
                        }
                    },
                },
            },
        )
        @inject
        async def renew_user(
            user_id: str,
            user_request: UserRequest,
            user_service: UserService = self.dependencies[0],
        ) -> JSONResponse:
            domain_user = UserMapper.to_domain(user_request)
            returned_user = await user_service.replace_user(user_id, domain_user)
            user_response = UserMapper.to_response(returned_user)
            response = JSONResponse(
                content=jsonable_encoder(user_response),
                status_code=status.HTTP_200_OK,
            )
            return response

        @APIRouter.api_route(
            self,
            path="/{user_id}",
            methods=["DELETE"],
            tags=["users"],
            description="API endpoint used to delete a user by its ID",
            responses={
                status.HTTP_200_OK: {
                    "model": UserResponse,
                    "description": "OK",
                    "content": {
                        "application/json": {
                            "example": {
                                "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
                                "name": "name",
                                "email": "email@email.com",
                                "created_at": "XXXX-XX-XXTXX:XX:XX.XXXXXX",
                                "updated_at": "",
                            }
                        }
                    },
                },
                status.HTTP_404_NOT_FOUND: {
                    "model": APIErrorResponse,
                    "description": "Not Found",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "",
                                "details": {"context": "", "cause": ""},
                                "isOperational": True,
                            }
                        }
                    },
                },
                status.HTTP_500_INTERNAL_SERVER_ERROR: {
                    "model": APIErrorResponse,
                    "description": "Internal Server Error",
                    "content": {
                        "application/json": {
                            "example": {
                                "message": "Internal Server Error",
                                "details": {"context": "", "cause": ""},
                                "isOperational": False,
                            }
                        }
                    },
                },
            },
        )
        @inject
        async def destroy_user(
            user_id: str,
            user_service: UserService = self.dependencies[0],
        ) -> JSONResponse:
            returned_user = await user_service.remove_user(user_id)
            user_response = UserMapper.to_response(returned_user)
            response = JSONResponse(
                content=jsonable_encoder(user_response),
                status_code=status.HTTP_200_OK,
            )
            return response

File: apps/server1/src/api/components/user/user-controller.ts
import express from 'express'
import httpStatus from 'http-status'
import {
  Body,
  Controller,
  Delete,
  Example,
  Get,
  Middlewares,
  Path,
  Post,
  Put,
  Query,
  Request,
  Response,
  Route,
  SuccessResponse,
  Tags,
} from 'tsoa'
import { inject, injectable } from 'tsyringe'

import { IAPIPaginationService } from '../../../services'
import {
  APIErrorResponse,
  APIPaginationResponse,
  validationMiddleware,
} from '../../shared'
import { UserMapper } from './user-mapper'
import { IUserService } from './user-service'
import { User, UserRequest, UserResponse } from './user-types'
import { userSchema } from './user-validator'

@injectable()
@Route('users')
@Tags('users')
class UserController extends Controller {
  constructor(
    @inject('UserService') private userService: IUserService,
    @inject('APIPaginationService')
    private apiPaginationService: IAPIPaginationService,
  ) {
    super()
  }

  /**
   * API endpoint used to create a new user.
   */
  @Post('/')
  @SuccessResponse('201', 'Created')
  @Example<UserResponse>({
    id: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX',
    name: 'name',
    email: 'email@email.com',
    createdAt: new Date(),
    updatedAt: null,
  })
  @Response<APIErrorResponse>('422', 'Unprocessable Entity', {
    message: 'Unprocessable Entity',
    detail: { context: 'context', cause: 'cause' },
    isOperational: true,
  })
  @Response<APIErrorResponse>('500', 'Internal Server Error', {
    message: 'Internal Server Error',
    detail: { context: 'context', cause: 'cause' },
    isOperational: false,
  })
  @Middlewares(validationMiddleware(userSchema))
  public async addUser(@Body() body: UserRequest): Promise<UserResponse> {
    const userMapper = new UserMapper()
    const user = userMapper.toDomain(body)
    const registeredUser = await this.userService.registerUser(user)
    const userResponse = userMapper.toResponse(registeredUser)
    this.setStatus(httpStatus.CREATED)
    return userResponse
  }

  /**
   * API endpoint used to get users through page-based pagination schema.
   * @param page The number of the page. If isn't provided, it will be set to 1.
   * @param limit The number of records per page. If isn't provided, it will be set to 1.
   */
  @Get('/')
  @SuccessResponse('200', 'OK')
  @Example<APIPaginationResponse<UserResponse>>({
    page: 1,
    limit: 1,
    totalPages: 1,
    totalRecords: 1,
    records: [
      {
        id: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX',
        name: 'name',
        email: 'email@email.com',
        createdAt: new Date(),
        updatedAt: null,
      },
    ],
    previous: 'previous',
    next: 'next',
  })
  @Response<APIErrorResponse>('500', 'Internal Server Error', {
    message: 'Internal Server Error',
    detail: { context: 'context', cause: 'cause' },
    isOperational: false,
  })
  public async fetchPaginatedUsers(
    @Request() req: express.Request,
    @Query() page?: number,
    @Query() limit?: number,
  ): Promise<APIPaginationResponse<UserResponse>> {
    const baseURL = `${req.protocol}://${req.get('host')}${req.originalUrl}`
    page = page ? page : 1
    limit = limit ? limit : 1
    const [retrievedUsers, totalRecords] =
      await this.userService.retrieveAndCountUsers(page, limit)
    const apiPaginationData = {
      page: page,
      limit: limit,
      totalRecords: totalRecords,
      records: retrievedUsers,
    }
    const apiPaginationResponse =
      this.apiPaginationService.createResponse<User>(baseURL, apiPaginationData)
    this.setStatus(httpStatus.OK)
    return apiPaginationResponse
  }

  /**
   * API endpoint used to get a user by its ID.
   */
  @Get('{userId}')
  @SuccessResponse('200', 'OK')
  @Example<UserResponse>({
    id: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX',
    name: 'name',
    email: 'email@email.com',
    createdAt: new Date(),
    updatedAt: null,
  })
  @Response<APIErrorResponse>('404', 'Not Found', {
    message: 'Not Found',
    detail: { context: 'context', cause: 'cause' },
    isOperational: true,
  })
  @Response<APIErrorResponse>('500', 'Internal Server Error', {
    message: 'Internal Server Error',
    detail: { context: 'context', cause: 'cause' },
    isOperational: false,
  })
  public async fetchUser(@Path() userId: string): Promise<UserResponse> {
    const retrievedUser = await this.userService.retrieveUser(userId)
    const userMapper = new UserMapper()
    const userResponse = userMapper.toResponse(retrievedUser)
    this.setStatus(httpStatus.OK)
    return userResponse
  }

  /**
   * API endpoint used to update a user by its ID.
   */
  @Put('{userId}')
  @SuccessResponse('200', 'OK')
  @Example<UserResponse>({
    id: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX',
    name: 'name',
    email: 'email@email.com',
    createdAt: new Date(),
    updatedAt: new Date(),
  })
  @Response<APIErrorResponse>('404', 'Not Found', {
    message: 'Not Found',
    detail: { context: 'context', cause: 'cause' },
    isOperational: true,
  })
  @Response<APIErrorResponse>('422', 'Unprocessable Entity', {
    message: 'Unprocessable Entity',
    detail: { context: 'context', cause: 'cause' },
    isOperational: true,
  })
  @Response<APIErrorResponse>('500', 'Internal Server Error', {
    message: 'Internal Server Error',
    detail: { context: 'context', cause: 'cause' },
    isOperational: false,
  })
  @Middlewares(validationMiddleware(userSchema))
  public async renewUser(
    @Path() userId: string,
    @Body() body: UserRequest,
  ): Promise<UserResponse> {
    const userMapper = new UserMapper()
    const user = userMapper.toDomain(body)
    const replacedUser = await this.userService.replaceUser(userId, user)
    const userResponse = userMapper.toResponse(replacedUser)
    this.setStatus(httpStatus.OK)
    return userResponse
  }

  /**
   * API endpoint used to delete a user by its ID.
   */
  @Delete('{userId}')
  @SuccessResponse('200', 'OK')
  @Example<UserResponse>({
    id: 'XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX',
    name: 'name',
    email: 'email@email.com',
    createdAt: new Date(),
    updatedAt: null,
  })
  @Response<APIErrorResponse>('404', 'Not Found', {
    message: 'Not Found',
    detail: { context: 'context', cause: 'cause' },
    isOperational: true,
  })
  @Response<APIErrorResponse>('500', 'Internal Server Error', {
    message: 'Internal Server Error',
    detail: { context: 'context', cause: 'cause' },
    isOperational: false,
  })
  public async destroyUser(@Path() userId: string): Promise<UserResponse> {
    const deletedUser = await this.userService.removeUser(userId)
    const userMapper = new UserMapper()
    const userResponse = userMapper.toResponse(deletedUser)
    this.setStatus(httpStatus.OK)
    return userResponse
  }
}

export { UserController }
